"Not checked" issue when error passed as arg

File `t_test.go`:
```go
package t

import (
	"encoding/json"
	"fmt"
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
)

type Foo struct {
	X interface{}
}

func TestSomething(t *testing.T) {
	// 1) This provokes a "error return value not checked" error
	assert.NoError(t, json.NewEncoder(os.Stdout).Encode(Foo{""}))

	// 2) This does not
	err := json.NewEncoder(os.Stdout).Encode(Foo{""})
	fmt.Println(err)
}
```

Checking it:
```shellsession
$ ./errchkjson t_test.go
.../t_test.go:18:20: Error return value of `(*encoding/json.Encoder).Encode` is not checked: unsafe type `interface{}` found
$ ./errchkjson -V
errchkjson version 0.2.3
commit: 3aae21891fb6867e0ae247403f8eb67c0573c5d5
built at: 2022-02-08T16:17:38Z
module version: v0.2.3, checksum: h1:97eGTmR/w0paL2SwfRPI1jaAZHaH/fXnxWTw2eEIqE0=
goos: linux
goarch: amd64
```

I find it kind of inconsistent that passing the error value from `Encode` here to another function -- in this case the often used `assert.NoError` -- provokes a linter failure (see 1) above), whereas simply capturing an error to a non-underscore variable does not (see 2)). (Ironically, for this particular example, 1) actually does check it due to nature of `assert.NoError`, whereas 2) doesn't, but that's a sidetrack.)

In my opinion, it would be prudent to skip reporting the error in case 1), i.e. when the returned err value is passed to some other function, because the error might actually be checked there. Just as in 2), capturing the error to a non-underscore variable means it could be checked later. As long as the tool does not make a deeper attempt to see if the error is _actually_ checked later on, I'm not sure why it would differentiate between capturing and passing.

--- a/errchkjson.go
+++ b/errchkjson.go
@@ -59,11 +59,11 @@ func (e *errchkjson) run(pass *analysis.Pass) (interface{}, error) {
 
 				switch fn.FullName() {
 				case "encoding/json.Marshal", "encoding/json.MarshalIndent":
-					e.handleJSONMarshal(pass, ce, fn.FullName(), true)
+					e.handleJSONMarshal(pass, ce, fn.FullName(), blankIdentifier)
 				case "(*encoding/json.Encoder).Encode":
-					e.handleJSONMarshal(pass, ce, fn.FullName(), true)
+					e.handleJSONMarshal(pass, ce, fn.FullName(), blankIdentifier)
 				default:
-					return true
+					e.inspectArgs(pass, ce.Args)
 				}
 				return false
 			}
@@ -85,9 +85,9 @@ func (e *errchkjson) run(pass *analysis.Pass) (interface{}, error) {
 
 			switch fn.FullName() {
 			case "encoding/json.Marshal", "encoding/json.MarshalIndent":
-				e.handleJSONMarshal(pass, ce, fn.FullName(), blankIdentifier(as.Lhs[1]))
+				e.handleJSONMarshal(pass, ce, fn.FullName(), evaluateMarshalErrorTarget(as.Lhs[1]))
 			case "(*encoding/json.Encoder).Encode":
-				e.handleJSONMarshal(pass, ce, fn.FullName(), blankIdentifier(as.Lhs[0]))
+				e.handleJSONMarshal(pass, ce, fn.FullName(), evaluateMarshalErrorTarget(as.Lhs[0]))
 			default:
 				return true
 			}
@@ -98,20 +98,28 @@ func (e *errchkjson) run(pass *analysis.Pass) (interface{}, error) {
 	return nil, nil
 }
 
-func blankIdentifier(n ast.Expr) bool {
+func evaluateMarshalErrorTarget(n ast.Expr) marshalErrorTarget {
 	if errIdent, ok := n.(*ast.Ident); ok {
 		if errIdent.Name == "_" {
-			return true
+			return blankIdentifier
 		}
 	}
-	return false
+	return variableAssignment
 }
 
-func (e *errchkjson) handleJSONMarshal(pass *analysis.Pass, ce *ast.CallExpr, fnName string, blankIdentifier bool) {
+type marshalErrorTarget int
+
+const (
+	blankIdentifier    = iota // the returned error from the JSON marshal function is assigned to the blank identifier "_".
+	variableAssignment        // the returned error from the JSON marshal function is assigned to a variable.
+	functionArgument          // the returned error from the JSON marshal function is passed to an other function as argument.
+)
+
+func (e *errchkjson) handleJSONMarshal(pass *analysis.Pass, ce *ast.CallExpr, fnName string, errorTarget marshalErrorTarget) {
 	t := pass.TypesInfo.TypeOf(ce.Args[0])
 	if t == nil {
 		// Not sure, if this is at all possible
-		if blankIdentifier {
+		if errorTarget == blankIdentifier {
 			pass.Reportf(ce.Pos(), "Type of argument to `%s` could not be evaluated and error return value is not checked", fnName)
 		}
 		return
@@ -131,15 +139,15 @@ func (e *errchkjson) handleJSONMarshal(pass *analysis.Pass, ce *ast.CallExpr, fn
 			pass.Reportf(ce.Pos(), "Error argument passed to `%s` does not contain any exported field", fnName)
 		}
 		// Only care about unsafe types if they are assigned to the blank identifier.
-		if blankIdentifier {
+		if errorTarget == blankIdentifier {
 			pass.Reportf(ce.Pos(), "Error return value of `%s` is not checked: %v", fnName, err)
 		}
 	}
-	if err == nil && !blankIdentifier && !e.omitSafe {
+	if err == nil && errorTarget == variableAssignment && !e.omitSafe {
 		pass.Reportf(ce.Pos(), "Error return value of `%s` is checked but passed argument is safe", fnName)
 	}
 	// Report an error, if err for json.Marshal is not checked and safe types are omitted
-	if err == nil && blankIdentifier && e.omitSafe {
+	if err == nil && errorTarget == blankIdentifier && e.omitSafe {
 		pass.Reportf(ce.Pos(), "Error return value of `%s` is not checked", fnName)
 	}
 }
@@ -266,6 +274,36 @@ func jsonSafeMapKey(t types.Type) error {
 	default:
 		// E.g. struct composed solely of basic types, that are comparable
 		return newUnsupportedError(fmt.Errorf("unsupported type `%s` as map key found", t.String()))
+	}
+}
+
+func (e *errchkjson) inspectArgs(pass *analysis.Pass, args []ast.Expr) {
+	for _, a := range args {
+		ast.Inspect(a, func(n ast.Node) bool {
+			if n == nil {
+				return true
+			}
+
+			ce, ok := n.(*ast.CallExpr)
+			if !ok {
+				return false
+			}
+
+			fn, _ := typeutil.Callee(pass.TypesInfo, ce).(*types.Func)
+			if fn == nil {
+				return true
+			}
+
+			switch fn.FullName() {
+			case "encoding/json.Marshal", "encoding/json.MarshalIndent":
+				e.handleJSONMarshal(pass, ce, fn.FullName(), functionArgument)
+			case "(*encoding/json.Encoder).Encode":
+				e.handleJSONMarshal(pass, ce, fn.FullName(), functionArgument)
+			default:
+				e.inspectArgs(pass, ce.Args)
+			}
+			return false
+		})
 	}
 }
 

--- a/testdata/src/nosafe/a.go
+++ b/testdata/src/nosafe/a.go
@@ -33,6 +33,7 @@ func JSONMarshalSafeTypesWithNoSafe() {
 	json.Marshal(nil)          // want "Error return value of `encoding/json.Marshal` is not checked"
 	_, err = json.Marshal(nil) // nil is safe, but omit-safe is set
 	_ = err
+	fmt.Print(json.Marshal(nil)) // nil is safe, error is passed as argument
 
 	_, _ = json.MarshalIndent(nil, "", " ")   // want "Error return value of `encoding/json.MarshalIndent` is not checked"
 	json.MarshalIndent(nil, "", " ")          // want "Error return value of `encoding/json.MarshalIndent` is not checked"
@@ -409,6 +410,7 @@ func JSONMarshalUnsafeTypes() {
 	_, _ = json.Marshal(f32)   // want "Error return value of `encoding/json.Marshal` is not checked: unsafe type `float32` found"
 	_, err = json.Marshal(f32) // err is checked
 	_ = err
+	fmt.Print(json.Marshal(f32)) // err is passed and therefore considered as checked
 
 	var f64 float64
 	_, _ = json.Marshal(f64)   // want "Error return value of `encoding/json.Marshal` is not checked: unsafe type `float64` found"
@@ -543,6 +545,7 @@ func JSONMarshalInvalidTypes() {
 	_, _ = json.Marshal(c64)   // want "`encoding/json.Marshal` for unsupported type `complex64` found"
 	_, err = json.Marshal(c64) // want "`encoding/json.Marshal` for unsupported type `complex64` found"
 	_ = err
+	fmt.Print(json.Marshal(c64)) // want "`encoding/json.Marshal` for unsupported type `complex64` found"
 
 	var c128 complex128
 	_, _ = json.Marshal(c128)   // want "`encoding/json.Marshal` for unsupported type `complex128` found"

--- a/testdata/src/standard/a.go
+++ b/testdata/src/standard/a.go
@@ -33,6 +33,7 @@ func JSONMarshalSafeTypes() {
 	json.Marshal(nil)          // nil is safe
 	_, err = json.Marshal(nil) // want "Error return value of `encoding/json.Marshal` is checked but passed argument is safe"
 	_ = err
+	fmt.Print(json.Marshal(nil)) // nil is safe, error is passed as argument
 
 	_, _ = json.MarshalIndent(nil, "", " ")   // nil is safe
 	json.MarshalIndent(nil, "", " ")          // nil is safe
@@ -409,6 +410,7 @@ func JSONMarshalUnsafeTypes() {
 	_, _ = json.Marshal(f32)   // want "Error return value of `encoding/json.Marshal` is not checked: unsafe type `float32` found"
 	_, err = json.Marshal(f32) // err is checked
 	_ = err
+	fmt.Print(json.Marshal(f32)) // err is passed and therefore considered as checked
 
 	var f64 float64
 	_, _ = json.Marshal(f64)   // want "Error return value of `encoding/json.Marshal` is not checked: unsafe type `float64` found"
@@ -543,6 +545,7 @@ func JSONMarshalInvalidTypes() {
 	_, _ = json.Marshal(c64)   // want "`encoding/json.Marshal` for unsupported type `complex64` found"
 	_, err = json.Marshal(c64) // want "`encoding/json.Marshal` for unsupported type `complex64` found"
 	_ = err
+	fmt.Print(json.Marshal(c64)) // want "`encoding/json.Marshal` for unsupported type `complex64` found"
 
 	var c128 complex128
 	_, _ = json.Marshal(c128)   // want "`encoding/json.Marshal` for unsupported type `complex128` found"

